src/main/java/org/eclipse/uprotocol/mqtt/HiveMqMQTT5Client.java

```diff
@@ -0,0 +1,288 @@
+        return MqttTopicFilter.builder()
+                .addLevel(String.valueOf(determinateClientIdentifierFromSource(source)))
+
+                //if source is null or predefined wildcard -> choose singleLevelWildcardAsString, otherwise choose value (the code is the other way around)
+                .addLevel(Optional.ofNullable(source).filter(_source -> !"*".equals(_source.getAuthorityName())).map(UUri::getAuthorityName).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(source).filter(_source -> _source.getUeId() != 0xffff).map(existingSource -> String.format("%04x", existingSource.getUeId())).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(source).filter(_source -> _source.getUeVersionMajor() != 0xff).map(UUri::getUeVersionMajor).map(Object::toString).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(source).filter(_source -> _source.getResourceId() != 0xffff).map(UUri::getResourceId).map(i -> String.format("%04x", i)).orElse(singleLevelWildcardAsString))
+
+                //if sink is null or predefined wildcard -> choose singleLevelWildcardAsString, otherwise choose value (the code is the other way around)
+                .addLevel(Optional.ofNullable(sink).filter(_sink -> !"*".equals(_sink.getAuthorityName())).map(UUri::getAuthorityName).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(sink).filter(_sink -> _sink.getUeId() != 0xffff).map(existingSource -> String.format("%04x", existingSource.getUeId())).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(sink).filter(_sink -> _sink.getUeVersionMajor() != 0xff).map(UUri::getUeVersionMajor).map(Object::toString).orElse(singleLevelWildcardAsString))
+                .addLevel(Optional.ofNullable(sink).filter(_sink -> _sink.getResourceId() != 0xffff).map(UUri::getResourceId).map(i -> String.format("%04x", i)).orElse(singleLevelWildcardAsString))
+
+                .build();
+    }
+
+    private char determinateClientIdentifierFromSource(UUri source) {
+        if (source == null) {
+            return MqttTopicFilter.SINGLE_LEVEL_WILDCARD;
+        }
+        if (source.getAuthorityName().equals("cloud")) { //Todo: fix determination of a cloud environment
+            return 'c';
+        }
+        return 'd';
+    }
+
+    private UAttributes extractUAttributesFromReceivedMQTTMessage(@NotNull Mqtt5Publish mqtt5Publish) {
+        if (mqtt5Publish.getTopic().getLevels().size() != 9)
+            throw new IllegalArgumentException("Topic did not match uProtocol pattern for mqtt messages of this spec");
+
+        Map<String, String> userProperties = convertUserPropertiesToMap(mqtt5Publish.getUserProperties());
+        UAttributes.Builder builder = UAttributes.newBuilder();
+
+        builder
+                .setSource(UUri.newBuilder()
```

Comment:
I think you can get the source and sink from the uAttributes in the user properties now (since the source and sink uuri are serialized) so do not need to parse the topic

Reply:
done